### PR TITLE
Data bindings

### PR DESCRIPTION
**Pull request by ghost**
First step for data bindings, with all the store and iterator logic.

They don't modify the layout yet, and are mostly untested.

### CODE REVIEW

File: examples/menu.markup [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
+        Welcome {{player.name}} to UIL!
+        <group class="btns-group">
+            <button class="btn-one-line" goto-view="play">Play</button>
+            <button class="btn-one-line" action="quit">Options</button>
+            <button class="btn" action="quit">Foo</button>
+            <button class="btn" action="quit">Bar</button>
+            <button class="btn-one-line" action="quit">More Options</button>
+            <button class="btn-one-line" action="quit">Quit</button>
+        </group>
+        <repeat iter="player" template-name="friend" />
```
Comment (Nemikolh) [inline examples/menu.markup]: `iter` should be `{{player.friends}}` and not `player`

File: examples/menu.markup [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
     </group>
 </view>
+
+<template name="friend">
+    {{player.name}} is {{player.status}}
```
Comment (Nemikolh) [inline examples/menu.markup]: here it should be `{{name}}` and `{{status}}` to be consistent with the doc.

File: parsers/src/markup/lib.rs [hunk 3]
```diff
@@ -66,36 +67,39 @@ impl<E> Library<E>
...
-            match test {
-                Some(found) => {
-                    mem::swap(
-                        child,
-                        &mut Node::from_template(
-                            found,
-                            NodeType::Group
-                        )
-                    );
+        let test = match node.node_type {
```
Comment (Nemikolh) [inline parsers/src/markup/lib.rs]: test should be renamed to node_type.

File: parsers/src/markup/tags.rs [hunk 4]
```diff
@@ -13,7 +13,7 @@ use uil_shared::markup::{
     LineInputData,
     ProgressBarData,
     TemplateData,
-    RepeatData
+    UnlinkedRepeatData
```
Comment (Nemikolh) [inline parsers/src/markup/tags.rs]: Why renaming it ? Keep the original naming.

**Comment by Nemikolh** [comment]
Okay lot's of work here ! ;)
So I jumped into renaming comments but found something new: `UnlinkedRepeat` tag.
What is that ?

File: parsers/src/markup/lib.rs [hunk 5]
```diff
@@ -66,36 +67,39 @@ impl<E> Library<E>
                                   templates: &HashMap<String, Template>,
                                   node: &mut Node)
     {
-        for child in node.children.iter_mut() {
-
-            let mut is_empty = None;
-            let test = match child.node_type {
```
Comment (Nemikolh) [inline parsers/src/markup/lib.rs:72]: Seems that the test naming was already there before ! :D

File: parsers/src/markup/lib.rs [hunk 6]
```diff
@@ -66,36 +67,39 @@ impl<E> Library<E>
...
+                match templates.get(template_name) {
+                    None => {
+                        err.log(format!(
+                                "Warning `{}` template name not found for repeat node",
+                                template_name));
+                        None
+                    }
+                    Some(found) => {
+                        let mut child = Node::from_template(found, NodeType::Group);
+                        replace_iterator_name(iter, &mut child);
```
Comment (Nemikolh) [inline parsers/src/markup/lib.rs]: This code is wrong, the way the tree is changed is incorrect.
This function should not change the tree that way, the Repeat tag should be handled at the view level.

**Comment by Nemikolh** [comment]
Okay I'll stop commenting and doing the change myself, there's too much factoring needed.
That's ok, the PR is huge and there's almost zero part of the code which is in a stable state and the comments are missing.

File: parsers/src/markup/parser.rs [hunk 7]
```diff
@@ -227,11 +227,13 @@ impl<E, B> Parser<E, B>
                 Some('{') => {
                     match buf_consumer.consume_any_char() {
                         Some('{') => {
+                            buf_consumer.consume_any_char();    // XXX: If I don't do this the data binding will contain one '{'
                             let mut data_binding = buf_consumer.consume_while(|c| c != '}').unwrap();
                             match buf_consumer.consume_any_char() {
                                 Some('}') => {
                                     match buf_consumer.consume_any_char() {
                                         Some('}') => {
+                                            buf_consumer.consume_any_char();    // XXX: If I don't do this the text after data binding will contain one '}'
```
Comment (Nemikolh) [inline parsers/src/markup/parser.rs]: This is not needed anymore (same for the one above) after merging master into the branch.

File: shared/src/markup.rs [hunk 8]
```diff
@@ -4,12 +4,14 @@ use std::collections::HashSet;
...
     Text(String),
     Binding(String),
+    RepeatBinding(RepeatBindingData),
     Group,
     Button(ButtonData),
     LineInput(LineInputData),
     ProgressBar(ProgressBarData),
     Template(TemplateData),
-    Repeat(RepeatData),
+    UnlinkedRepeat(UnlinkedRepeatData),
```
Comment (Nemikolh) [inline shared/src/markup.rs]: Those changes on Repeat should be reverted to what they were before. There should be just one enum value as before and not three.

**Comment by lthms** [comment]
Looks like you work a lot here (even if tests seem broken).

Can you explain where you are and what is still left to do?

**Comment by Nemikolh** [comment]
I have simplified a lot the original work we did to keep things simple. I also tried to improve the documentation and the naming conventions. I'm now updating the tests for the new crate `oil_databindings` and I will then upgrade the root crate to use it correctly. Once this is done, I will merge this branch into master before moving to the `fonts` branch to add the `fonts` support.

By the way, this branch also include gamma correction and numerous fixes which have nothing to do with `data_bindings`.

**Comment by lthms** [comment]
That sounds pretty great! Do you have simple yet useful tasks I can do to help?

**Comment by Nemikolh** [comment]
If you want to go and start a PR/discussion for the user events system (#15) I would gladly accept your help ! :smiley:

**Comment by lthms** [comment]
Congrats'!

**Comment by Nemikolh** [comment]
Thanks ! :smiley_cat:

**Pull request merged**